### PR TITLE
Split into view, viewmodel, and factorio-blueprint

### LINKED ISSUE #16: Should we refactor to decouple logic and view?
Right now there's some heavy coupling, which makes the code harder to reason about. Anyone up for an attempt at some restructuring? Any comments about stuff that should be taken into consideration, etc?

### LINKED ISSUE #36: click and drag does not place in first cell
When clicking and dragging to place items, the item is not placed in the first cell that was clicked. Also happens with removing items.

### LINKED ISSUE #39: Underground Belt Rotation
This sorta has to do with #8, Currently we have two icons for input/output of each belt. However they are all facing the same direction so having properly connected belts will make it look something like
this 
It would be good if at the very least had the  belts mirrored when rotated so that they go the same direction without being upside down.
Would it be possible to take the sprite sheet, cut out the needed ones and cycle through the individual rotations instead of having a single texture?

### LINKED ISSUE #42: Refactor Placeable Variable into Objects instead of Multidimensional Array

### PR DESCRIPTION
**Pull request by ryepup**
Refactored to break apart the logic, and use https://github.com/demipixel/factorio-blueprint/ as the data model.

* script.js - run app-startup logic, mostly hooking up DOM to view
* view.js - functions for building various UI elements, and reacting to different UI and model events
* viewmodel.js - functions for managing the state of the application, a mix of synchronous calls and event-driven callbacks

The view changes the original implementation significantly; I think it's doing more DOM operations, and less CSS positioning.

A lot of this is pretty clumsy, written in an old-school style to support IE11 without any build process. If #45 goes through then much of this code could be better expressed as a handful of classes.

There are still bugs in here, but they seem to match the ones in `master` so I think this refactor is close enough. Several issues will get much easier if this gets merged: #25, #5, #4. We should also get fewer merge conflicts moving forward, since the code isn't all in one place.

fixes #16
fixes #39
fixes #42
fixes #36

### CODE REVIEW

File: css/style.css [hunk 1]
```diff
@@ -72,6 +72,10 @@ html {
     opacity: 0.75;
 }
 
+.mouse__preview.blocked {
```
Comment (Zeragamba) [inline css/style.css:75]: wouldn't `blocked` be a modifier on the preview? eg. `.mouse__preview--blocked`

not 100% sure how BEM works for this kind of thing.

File: js/viewmodel.js [hunk 2]
```diff
@@ -65,7 +138,7 @@ window.FBE = window.FBE || {};
 
     function addDerivedProperties(entity) {
 
-        var nonRotators = /beacon|roboport|lab|heat_pipe|pipe|furnace|chest|pole|substation|solar|accumulator|centrifuge|wall|rocket_silo|radar|turret|speaker|power_switch/i;
+        var nonRotators = /beacon|roboport|lab|heat_pipe|pipe|furnace|chest|pole|substation|solar|accumulator|centrifuge|wall|rocket_silo|radar|turret|speaker|power_switch|reactor|lamp|land_mine/i;
```
Comment (SpyMaster356) [inline js/viewmodel.js]: well, that be a piece of regex...

wouldn't an array and `#indexOf` work just as well?

File: js/viewmodel.js [hunk 3]
```diff
@@ -79,7 +152,10 @@ window.FBE = window.FBE || {};
         var directionPrefix = entity.directionType
             ? entity.type[0] + '-'
             : '';
-        var fileName = directionPrefix + entity.name.replace(/_/g, '-') + '.png';
+
+        var fileName = directionPrefix
+            + entity.name.replace(/_/g, '-') // TODO: rename files so we can drop this
```
Comment (SpyMaster356) [inline js/viewmodel.js]: opened #49 to get this fixed

File: js/view.js [hunk 4]
```diff
@@ -143,9 +152,17 @@ window.FBE = window.FBE || {};
         div.classList.add('preview');
         div.style.width = placeable.width * 32 - 2 + "px";
         div.style.height = placeable.height * 32 - 2 + "px";
-        if (placeable.direction) {
-            // TODO account for mismatch with icons rotation and CSS rotation
-            div.style.transform = 'rotate(' + 45 * placeable.direction + 'deg)';
+        if (placeable.canRotate) {
+            var rotation = (placeable.direction - placeable.defaultDirection);
```
Comment (SpyMaster356) [inline js/view.js]: seems like the brackets are unneeded

File: js/viewmodel.js [hunk 5]
```diff
@@ -200,4 +203,13 @@ window.FBE = window.FBE || {};
         return 'vendor/factorio/icons/placeable/' + fileName;
     }
 
+    function getDefaultDirection(entity) {
+        if (entity.type === 'input') { return 2; }
+        if (entity.type === 'output') { return 6; }
+        if (/inserter|offshore_pump/.test(entity.name)) { return 6; }
```
Comment (SpyMaster356) [inline js/viewmodel.js]: can we move these regexes out into constants and/or arrays?

File: js/viewmodel.js [hunk 6]
```diff
@@ -44,7 +44,7 @@ window.FBE = window.FBE || {};
 
     function loadEntity(placeables, entity) {
         var items = placeables
-            .filter(function (p) { return p.name === entity.name; }),
+                .filter(function (p) { return p.name === entity.name; }),
```
Comment (Zeragamba) [inline js/viewmodel.js]: can we move this onto the previous line, or give `item` there it's own `var` statement.

File: js/viewmodel.js [hunk 7]
```diff
@@ -157,7 +157,12 @@ window.FBE = window.FBE || {};
                     },
                     rawEntities[name],
                     // fix bad data in factorio-blueprint
-                    name === 'steam_turbine' && {width: 5, height: 3}                 
+                    name === 'steam_turbine' && { width: 5, height: 3 },
```
Comment (SpyMaster356) [inline js/viewmodel.js]: can we move these into a switch statement? I found this really hard to understand what it was doing.

File: package.json [hunk 8]
```diff
@@ -1,11 +1,11 @@
 {
   "name": "factorio_blueprint_editor",
-  "version": "1.0.0",
+  "version": "1.1.0",
   "description": "Online Editor for Elxeno's Factorio Blueprint Editor",
   "main": "index.js",
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "lint": "./node_modules/.bin/eslint js/**/*.js"
+    "lint": "eslint js/**/*.js"
```
Comment (Zeragamba) [inline package.json:8]: we should keep using the project's version of eslint instead of relying on a globally specified version

**Review by Zeragamba — CHANGES_REQUESTED** [review]

File: js/viewmodel.js [hunk 9]
```diff
@@ -0,0 +1,221 @@
+window.FBE = window.FBE || {};
+
+(function (FBE) {
+    'use strict';
+    /* global Blueprint, TinyEmitter */
+
+    var events = new TinyEmitter(),
+        bp = new Blueprint(),
+        selectedItem,
+        entityNamesThatDoNotRotate = /beacon|roboport|lab|heat_pipe|pipe$|furnace|chest|pole|substation|solar|accumulator|centrifuge|wall|rocket_silo|radar|turret|speaker|power_switch|reactor|lamp|land_mine/i
```
Comment (Zeragamba) [inline js/viewmodel.js]: well, that be a piece of regex...

wouldn't an array and #indexOf work just as well?

File: js/viewmodel.js [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+        });
+    }
+
+    function getIconUrl(entity) {
+        var directionPrefix = entity.directionType
+            ? entity.type[0] + '-'
+            : '';
+
+        var fileName = directionPrefix
+            + entity.name.replace(/_/g, '-') // TODO: rename files so we can drop this
```
Comment (Zeragamba) [inline js/viewmodel.js:220]: opened #49 to get this fixed

File: js/viewmodel.js [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+            + '.png';
+        return 'vendor/factorio/icons/placeable/' + fileName;
+    }
+
+    function getDefaultDirection(entity) {
+        if (entity.type === 'input') { return 2; }
+        if (entity.type === 'output') { return 6; }
+        if (/inserter|offshore_pump/.test(entity.name)) { return 6; }
```
Comment (Zeragamba) [inline js/viewmodel.js]: can we move these regexes out into constants and/or arrays?

File: js/viewmodel.js [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+                        name: name
+                    },
+                    rawEntities[name],
+                    // fix bad data in factorio-blueprint
+                    name === 'steam_turbine' && { width: 5, height: 3 },
+                    name === 'boiler' && { width: 3, height: 2 },
+                    // fix spots where height/width are swapped
+                    name === 'pump' && { width: 2, height: 1 },
+                    name === 'decider_combinator' && { width: 2, height: 1 },
+                    name === 'arithmetic_combinator' && { width: 2, height: 1 }
```
Comment (Zeragamba) [inline js/viewmodel.js]: can we move these into a switch statement? It took me a while to understand what these && were doing

File: js/view.js [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        }
+    }
+
+    function createPreview(placeable) {
+        var div = document.createElement("div");
+        div.classList.add('preview');
+        div.style.width = placeable.width * 32 - 2 + "px";
+        div.style.height = placeable.height * 32 - 2 + "px";
+        if (placeable.canRotate) {
+            var rotation = (placeable.direction - placeable.defaultDirection);
```
Comment (Zeragamba) [inline js/view.js]: seems like the brackets are unneeded

**Comment by Zeragamba** [comment]
I apologize for the email spam...

File: js/viewmodel.js (hunk 12 above)
Comment (ryepup) [reply js/viewmodel.js]: will do

File: js/view.js (hunk 13 above)
Comment (ryepup) [reply js/view.js]: agreed; at some point I was trying to work it out with `%` but that was a dead end.

File: package.json (hunk 8 above)
Comment (ryepup) [reply package.json:8]: npm handles this already, the `./node_modules/.bin` is redundant.

It basically puts `./node_modules/.bin` at the front of your `PATH` so we are using project's dependencies before anything installed system wide.

The official docs are very terse on this: https://docs.npmjs.com/misc/scripts#path

File: js/viewmodel.js (hunk 9 above)
Comment (ryepup) [reply js/viewmodel.js]: it'd be a much bigger array. With the regex I can say `chest` , whereas an array would need `wooden_chest`, `iron_chest`, etc. similar story for turrets, poles, furnaces.

File: js/viewmodel.js (hunk 11 above)
Comment (ryepup) [reply js/viewmodel.js]: will do

File: package.json (hunk 8 above)
Comment (Zeragamba) [reply package.json:8]: Oh. I did not know that. Nvm then

File: js/viewmodel.js [hunk 14]
```diff
@@ -156,13 +179,7 @@ window.FBE = window.FBE || {};
...
                     },
                     rawEntities[name],
-                    // fix bad data in factorio-blueprint
-                    name === 'steam_turbine' && { width: 5, height: 3 },
-                    name === 'boiler' && { width: 3, height: 2 },
-                    // fix spots where height/width are swapped
-                    name === 'pump' && { width: 2, height: 1 },
-                    name === 'decider_combinator' && { width: 2, height: 1 },
-                    name === 'arithmetic_combinator' && { width: 2, height: 1 }
+                    factorioBlueprintOverrides[name]
```
Comment (Zeragamba) [inline js/viewmodel.js]: That's definitely clearer to read.

File: js/viewmodel.js [hunk 15]
```diff
@@ -7,7 +7,30 @@ window.FBE = window.FBE || {};
     var events = new TinyEmitter(),
         bp = new Blueprint(),
         selectedItem,
-        entityNamesThatDoNotRotate = /beacon|roboport|lab|heat_pipe|pipe$|furnace|chest|pole|substation|solar|accumulator|centrifuge|wall|rocket_silo|radar|turret|speaker|power_switch|reactor|lamp|land_mine/i
+        entityNamesThatDoNotRotate = /beacon|roboport|lab|heat_pipe|pipe$|furnace|chest|pole|substation|solar|accumulator|centrifuge|wall|rocket_silo|radar|turret|speaker|power_switch|reactor|lamp|land_mine/i,
```
Comment (Zeragamba) [inline js/viewmodel.js]: These regexes still feel wrong, as they are just a list of options. i think these would be better handled by an array.

Also doing a little bit of searching, i think this regex match performs a lot slower than using indexOf

File: js/viewmodel.js [hunk 16]
```diff
@@ -212,10 +229,11 @@ window.FBE = window.FBE || {};
...
         if (entity.type === 'output') { return 6; }
-        if (/inserter|offshore_pump/.test(entity.name)) { return 6; }
-        if (/splitter|burner_mining/.test(entity.name)) { return 4; }
-        if (/electric_mining|pipe_to_ground|steam_engine|steam_turbine|pump|gate|arithmetic_combinator|decider_combinator/.test(entity.name)) { return 2; }
-        return 0;
+
+        var first = defaultDirectionsByEntityName
+            .filter(function(x){ return x.regex.test(entity.name);})[0];
+
+        return (first && first.direction) || 0;
```
Comment (Zeragamba) [inline js/viewmodel.js]: still doing this strange conditional that returns a non-boolean

File: js/viewmodel.js [hunk 17]
```diff
@@ -7,7 +7,30 @@ window.FBE = window.FBE || {};
     var events = new TinyEmitter(),
         bp = new Blueprint(),
         selectedItem,
-        entityNamesThatDoNotRotate = /beacon|roboport|lab|heat_pipe|pipe$|furnace|chest|pole|substation|solar|accumulator|centrifuge|wall|rocket_silo|radar|turret|speaker|power_switch|reactor|lamp|land_mine/i
+        entityNamesThatDoNotRotate = /beacon|roboport|lab|heat_pipe|pipe$|furnace|chest|pole|substation|solar|accumulator|centrifuge|wall|rocket_silo|radar|turret|speaker|power_switch|reactor|lamp|land_mine/i,
+        defaultDirectionsByEntityName = [
```
Comment (Zeragamba) [inline js/viewmodel.js]: the variable name sounds like it will contain a map, with a EntityName being the key. `entityDefaultDirections` maybe?

**Review by Zeragamba — CHANGES_REQUESTED** [review]

File: js/viewmodel.js (hunk 15 above)
Comment (Danielv123) [reply js/viewmodel.js]: I believe Array.includes() is what you should be using to check the existence of a string in an array

File: js/viewmodel.js (hunk 16 above)
Comment (ryepup) [reply js/viewmodel.js]: Using the logical operators to for short-circuit evaluation is a pretty common idiom, and the simplest way to guard against nulls or provide default values. 

See also: http://eloquentjavascript.net/01_values.html#h_3jN0iK4yKW

In this case, I think a ternary will be clearer, but it'll still be using non-booleans / truthiness

File: css/style.css (hunk 1 above)
Comment (ryepup) [reply css/style.css:75]: I don't know BEM, is this project strict about BEM usage?

File: js/viewmodel.js (hunk 15 above)
Comment (ryepup) [reply js/viewmodel.js]: This regex is called once at app startup and whenever we load a blueprint. I don't think it's a performance bottleneck.

Profiling on my aging laptop (i5, chrome, ubunut) it takes 5ms to call `getPlaceableItems`, and of that, 0.16ms is spent in `entityNamesThatDoNotRotate.test`.

Is that fast enough?

File: js/viewmodel.js (hunk 17 above)
Comment (ryepup) [reply js/viewmodel.js]: will do

File: js/viewmodel.js (hunk 14 above)
Comment (ryepup) [reply js/viewmodel.js]: agreed

File: js/viewmodel.js (hunk 15 above)
Comment (Zeragamba) [reply js/viewmodel.js]: right, but this still feels like an abuse of regex. It's being used like an array, instead of using an array.

Comment (Zeragamba) [reply js/viewmodel.js]: Will merge this in for now, but opened #54 to get this fixed

**Review by Zeragamba — APPROVED** [review]

**Comment by Zeragamba** [comment]
@camerongillette can you resolve the merge conflicts please?

**Comment by ryepup** [comment]
resolved merge conflicts. With ES6 on the table there's a lot of cleanup that can be done; I was sorely missing arrow functions especially

**Review by Zeragamba — APPROVED** [review]

**Pull request merged**